ovos_plugin_manager/templates/language.py

```diff
@@ -1,28 +1,74 @@
+from json_database import JsonStorageXDG
+import xdg.BaseDirectory
+
 
 class LanguageDetector:
     def __init__(self, config=None):
         self.config = config or {}
+        db_name = self.__class__.__name__.lower()
+        self.cache = JsonStorageXDG(db_name,
+                                    xdg_folder=xdg.BaseDirectory.xdg_cache_home,
+                                    subfolder="lang_detections")
         self.default_language = self.config.get("lang") or "en-us"
         # hint_language: str  E.g., 'it' boosts Italian
         self.hint_language = self.config.get("hint_lang") or self.default_language
         # boost score for this language
         self.boost = self.config.get("boost")
 
+    def get_detection(self, text):
+        # plugins that need/want cache can override this method
+        # old plugins (pre-cache) or if cache is not desired
+        # can continue overriding self.detect and will remain working as usual
+        # usage of self.get_detection is encouraged for non-local providers
+        return self.default_language  # assume default language by default
+
     def detect(self, text):
-        # assume default language
-        return self.default_language
+        # read from cache
+        tx = self.cache.get(text)
+        if tx:
+            return tx
+        # do detection
+        tx = self.get_detection(text)
+        # cache translation
+        self.cache[text] = tx
+        self.cache.store()
+        return tx
 
     def detect_probs(self, text):
-        return {self.detect(text): 1}
+        return {self.detect(text): 1.0}
 
 
 class LanguageTranslator:
     def __init__(self, config=None):
         self.config = config or {}
+        db_name = self.__class__.__name__.lower() + ".json"
+        self.cache = JsonStorageXDG(db_name,
+                                    xdg_folder=xdg.BaseDirectory.xdg_cache_home,
+                                    subfolder="lang_translations")
         # translate from, unless specified/detected otherwise
         self.default_language = self.config.get("lang") or "en-us"
         # translate to
         self.internal_language = self.config.get("lang") or "en-us"
 
-    def translate(self, text, target=None, source=None):
+    def get_translation(self, text, target=None, source=None):
+        # plugins that need/want cache can override this method
```

Comment:
Consider turning this comment into a docstring

Reply:
maybe we should open an issue in existing plugins, at least the online ones

to benefit from cache those all need to change `def translate` to `def get_translation`

this should probably also be documented somewhere else, new plugins should be encouraged to always use `get_translation`

Reply:
ah, right.. Maybe `translate` should be refactored to `_translate` and `translate` calls the internal method and logs a deprecation warning saying to use `get_translation`?

Reply:
i think it makes sense to keept both, with the aim of having a explicit way to skip the cache

Reply:
could the cache be a decorator? it just seems like we shouldn't have 2 separate methods to override in the base classes, since the cache is the same in all of them